gnome-image-installer/util/gis-unattended-config.c

```diff
@@ -416,66 +418,77 @@ gis_unattended_config_match_computer (GisUnattendedConfig *self,
   return GIS_UNATTENDED_COMPUTER_DOES_NOT_MATCH;
 }
 
-static GFile *
-invent_backup_name (GFile *source)
+static gchar *
+invent_backup_name (const gchar *source,
+                    GError     **error)
 {
-  g_autoptr(GFile) dir = g_file_get_parent (source);
-  g_autofree gchar *basename = g_file_get_basename (source);
-  g_autofree gchar *uuid = g_uuid_string_random ();
+  g_autofree gchar *dirname = g_path_get_dirname (source);
+  g_autofree gchar *basename = g_path_get_basename (source);
   gchar *dot = strrchr (basename, '.');
   g_autofree gchar *target_basename = NULL;
+  g_autofree gchar *target = NULL;
+  gint fd = -1;
 
   if (dot == NULL)
     {
-      target_basename = g_strjoin (".", basename, uuid, NULL);
+      target_basename = g_strjoin (".", basename, "XXXXXX", NULL);
     }
   else
     {
       *dot = '\0';
-      target_basename = g_strjoin (".", basename, uuid, dot + 1, NULL);
+      target_basename = g_strjoin (".", basename, "XXXXXX", dot + 1, NULL);
     }
 
-  return g_file_get_child (dir, target_basename);
+  target = g_build_path (G_DIR_SEPARATOR_S, dirname, target_basename, NULL);
+  if ((fd = g_mkstemp (target)) < 0)
+    return glnx_null_throw_errno_prefix (error,
+                                         "Couldn't create a backup file in ‘%s’",
+                                         dirname);
+
+  /* We just want the fresh filename */
+  if (!g_close (fd, error))
+    return NULL;
+
+  return g_steal_pointer (&target);
 }
 
 static gboolean
-rename_existing_file_if_exists (GFile   *source,
-                                gchar  **backup_path,
-                                GError **error)
+key_file_save_to_file_with_backup (GKeyFile    *key_file,
+                                   const gchar *target_path,
+                                   gchar      **backup_basename,
+                                   GError     **error)
 {
-  g_autoptr(GFile) target = invent_backup_name (source);
-  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GFile) target_file = g_file_new_for_path (target_path);
+  g_autofree gchar *backup_path = NULL;
 
-  if (g_file_move (source, target,
-                   /* No G_FILE_COPY_OVERWRITE, so target will not be
-                    * overwritten if it already exists. Since its name includes
-                    * a fresh UUID, this is extremely unlikely.
-                    */
-                   G_FILE_COPY_NO_FALLBACK_FOR_MOVE,
-                   NULL, NULL, NULL, &local_error))
+  if (!g_file_query_exists (target_file, NULL))
     {
-      *backup_path = g_file_get_basename (target);
-      return TRUE;
+      *backup_basename = NULL;
+      return g_key_file_save_to_file (key_file, target_path, error);
```

Comment:
It’s sad that there’s still a TOCTTOU race here, but no way around it since `g_key_file_save_to_file()` unconditionally overwrites things. :-(